iriscasttools/iriscasttools/utils.py

```diff
@@ -0,0 +1,224 @@
+                        i + 1,
+                        seconds,
+                    )
+                time.sleep(seconds)
+            return res
+
+        return inner
+
+    return decorator
+
+
+@retry(retry_on=(AssertionError,), retries=3, delay=3, backoff=2, retry_logger=logger)
+def run_cmd(cmd_args: str):
+    """Run a bash command with given arguments and return output
+
+    Keyword Arguments
+        cmd_args: a string representing bash command to run
+    """
+    with subprocess.Popen(
+        cmd_args, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE
+    ) as out:
+        # assert we did not find any errors
+        assert not out.stderr.read().decode()
+        res_out = out.stdout.read().decode()
+    return res_out
+
+
+def check_ipmi_conn():
+    """Check if IPMItool exists and can be connected to
+
+    Checks if device exists at any of these locations /dev/ipmi0, /dev/ipmi/0 or /dev/ipmidev/0
+    which imply that ipmi-dcmi can be used to get power info
+    """
+    res = any(Path(f).exists() for f in ["/dev/ipmi0", "/dev/ipmi/0", "/dev/ipmidev/0"])
+    if not res:
+        logger.error("Failed to find ipmi device on host")
+    return res
+
+
+def ipmi_raw_power_query():
+    """
+    Get raw power query from IPMItool
+
+    Calls ipmi-dcmi command to get all power statistics
+    """
+    try:
+        return run_cmd("/usr/sbin/ipmi-dcmi --get-system-power-statistics")
+    except AssertionError:
+        logger.error(
+            "command '/usr/sbin/ipmi-dcmi --get-system-power-statistics failed"
+        )
+        return None
+
+
+def to_csv(stats: Dict, include_header: bool = False):
+    """convert dictionary into csv string
+
+    Keyword arguments
+    stats -- dict, key-value pairs to be returned in csv format
+    include_header, bool, flag to set if dictionary keys should be returned as header
+    """
+    res = ",".join([str(s) for s in stats.values()])
+    if include_header:
+        return f"{','.join(stats.keys())}\n{res}"
+    return res
+
+
+def get_ipmi_power_stats(*args):
+    """Get ipmi power stats as dictionary.
+
+    Keyword arguments
+    *args -- a set of fields to parse and collect.
+    can be one or more of:
+        "current_power": instantaneous reading in Watts,
+        "minimum_power_over_sampling_duration: minimum Watt reading from last sample till now
+        "maximum_power_over_sampling_duration": maximum Watt reading from last sample till now
+        "average_power_over_sampling_duration": average Watt reading from last sample till now
+        "statistics_reporting_time_period": sample period in milliseconds
+        "time_stamp": current timestamp for reading taken
+        "power_measurement": whether power measurement is active on host
+    """
+    power_stats = {x: "" for x in args}
+
+    if not check_ipmi_conn():
+        return power_stats
+    res = ipmi_raw_power_query()
+
+    if res:
```

Comment:
If `ipmi_raw_power_query` throws this whole block can be simplified by removing the if condition